IoHandle.md

```diff
@@ -0,0 +1,387 @@
+
+### Refactoring the FD
+
+
+##### Overview
+
+The issue with refactoring the fd is the implicit assumption that the main interface to Envoy from the application is a socket. In this PR, our purpose is to abstract away sockets and all system calls pertaining to sockets from the I/O interface. The IoHandle represents the abstract interface for all I/O operations and the details of any particular interface (socket, VPP, etc) are contained in classes derived from IoHandle. 
+To this end, we also define the IoAddress class and some derivatives, as well as the IoDescriptor class, which will replace the fd() calls when dealing with actual sockets.
+
+These changes in this PR are a proposal and are intended to get feedback for the direction we are taking. Rather than show all the changes required to implement this new direction, we will also discuss towards the end how we anticipate further changes rippling through the code.
+
+Note also that this PR does not address events yet, as previous comments suggested. The intention is to introduce further PRs that will address events and the dispatcher.
+
+
+
+###### IoHandle and derivative classes
+
+The following definitions would be contained in a new file io_defs.h  (envoy/include/envoy/network):
+
+```
+
+// Generalize the SysCallResult to accommodate I/O operations that are not
+// carried out by the kernel stack, hence not accomplished through a syscall
+
+template <typename T> struct IoCallResult : public struct IoResult  { 
+      T rc_;
+       int errno_;
+};
+typedef IoCallResult<int> IoCallIntResult;
+typedef IoCallResult<ssize_t> IoCallSizeResult;
+typedef IoCallResult<void*> IoCallPtrResult;
+
+template <typename T> class IoDescriptor  {
+public:
+   IoDescriptor(T descriptor):descriptor_(descriptor) { }
+   virtual ~IoDescriptor() { };
+
+   T descriptor() const { return(descriptor_); }
+
+   operator(T) const { return(descriptor_);
+   bool	operator == (const T& lhs) { return(descriptor_== lhs); }
+   bool	operator != (const T& lhs) { return(descriptor_!=lhs); }
+
+protected:
+   T descriptor_;
+};
+
+typedef IoDescriptor<int> IoSocketDescriptor;  // fd for sockets
+typedef IoDescriptor<uint32_t> IoVppDescriptor;  // for example…
```

Comment:
@alyssawilk: thank you for your comments. It's clear that this PR way overshoots what you had in mind. Our thinking was stated in the first few sentences of the PR, and it was my misunderstanding that we really need to get rid of the implicit socket interface in order to refactor the fd. What you propose is much narrower in scope and a lot less work- happy to do it :-)

Before I respond to the details of your comments, my plan is to produce a new PR ASAP that puts all I/O methods into the IoHandle class and it will be used accordingly to replace the existing fd() calls.

I'm not sure what you mean in stating "we're still allowing fd() returns for classes accessing IoDescriptor." The intention was that for a class derived from IoDescriptor, such as IoSocketDescriptor, would return the equivalent of an fd with the IoSocketDescriptor::descriptor() method.

As for how you see IoHandles for the different interfaces, I see what you have in mind and that's easy to do. For the socket derivatives of IoHandles, the handle() method would replace the fd() calls. Agree?

Reply:
So my personal preference would be that long term we have no raw fd access.  If someone (e.g. Google) has some extra functions they want on their sockets (e.g. setting internal ToS bits) we'd have

class IoHandle{};
class FdIoHandleImpl : virtual public IoHandle {};
class GoogleIoHandle { virtual void SetQoS() PURE; };
class GoogleFdIoHandleImpl : public FdIoHandleImpl, public GoogleIoHandle {}

which I think works from an inheritance standpoint, and also works for things like VPP and QUIC where you'd have to have custom implementations per impl type anyway.

My comment regarding having an fd() function was more about the migration/refactor strategy.  Rather than try to get all this done in one fell swoop, you could change the TransportSocket from returning an int fd() to returning an IoHandle* handle() where the IoHandle interface is literally just fd().  That lets you do one PR which does *all* the interface changes with a fairly easy sed regex where transport_socket->fd() -> transport_socket->io_handle()->fd()

We then envoy-announce that fd() access is going away.  Anyone with downstream builds which use raw fd access can start implementing their own MyCompanyIoHandles while we sort out the Envoy changes

The Envoy changes can then be a series of PRs where we replace calls which use the fd with clean APIs in IOHandle.  I think these may provoke a fair bit of discussion and debate about ideal interface.   Beyond giving folks time to change their own Envoy builds, I don't think you want to try do all the back and forth on prefecting the interface for every fd access call in one giant PR resolving merge conflicts as other folks add fd access :-P  It could be that we nail the API right on the first pass, but really large monolithic changes tend to be hard on reviewers and take a long time to land so I think it's worth splitting them up either way.

Finally once everything in Envoy is in IOHandle and no callers reference fd() (and we've given downstream folks time to do their own changes, (even if their Api is a MyIoHandleImpl with fd() access :-)) we remove fd() from the IoHandle API.

Does that make sense and sound sane?  I don't think it's much extra work, and I think it'll make the whole refactor go much more smoothly

Reply:
+1 to what @alyssawilk is saying above. This is/was my understanding of how we go about doing this work. I don't think it's worth it to introduce an intermediate IoDescriptor class when the end goal is to hide any concept of descriptors?

Reply:
@alyssawilk, @mattklein123: can you tell me how you would define the abstract class IoHandle then? In the discussions about the fd refactor, it has been said more than once that the ioHandle should contain all the socket methods. Do you see it this way? 

Understand that I am not arguing here :-), I'm just looking for clarity. I will follow the suggestions you make, but I'm trying to reconcile your comments with the previous discussions. So in order to avoid further churn, please tell me how you see the abstract class IoHandle. Thank you.

Reply:
Basically IoHandle should have all the platform specific methods that we currently use, such as bind(), read(), write(), setOption() (or setNoDelay() per my other comment), etc. Then there can be a platform implementation of IoHandle() which does the right thing. And in v1, if you spit this out of the dispatcher, like we do for connections currently, it should be pretty easy to hide the details IMO.